File: backend/app/routers/credit.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app import models, schemas
from app.database import get_db
from app.deps import get_current_household_id
from app.services.borrowing_capacity import compute_borrowing_capacity
from app.services.credit_advisor import build_recommendations, score_band
from app.services.finance_utils import LIABILITY_TYPES

router = APIRouter(prefix="/credit", tags=["credit"])


def _get_household_user_or_404(db: Session, household_id: str, user_id: str) -> models.User:
    user = (
        db.query(models.User)
        .filter(models.User.id == user_id, models.User.household_id == household_id)
        .first()
    )
    if not user:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Household member not found")
    return user


@router.get("/profiles", response_model=list[schemas.CreditProfileOut])
def list_profiles(db: Session = Depends(get_db), household_id: str = Depends(get_current_household_id)):
    return (
        db.query(models.CreditProfile)
        .join(models.User)
        .filter(models.User.household_id == household_id)
        .all()
    )


@router.put("/profiles/{user_id}", response_model=schemas.CreditProfileOut)
def upsert_profile(
    user_id: str,
    payload: schemas.CreditProfileUpsert,
    db: Session = Depends(get_db),
    household_id: str = Depends(get_current_household_id),
):
    _get_household_user_or_404(db, household_id, user_id)

    data = payload.model_dump(exclude={"recorded_date"})
    profile = db.query(models.CreditProfile).filter(models.CreditProfile.user_id == user_id).first()
    if profile:
        for field, value in data.items():
            setattr(profile, field, value)
    else:
        profile = models.CreditProfile(user_id=user_id, **data)
        db.add(profile)

    db.add(
        models.CreditScoreHistory(
            user_id=user_id,
            bureau=payload.bureau,
            score=payload.score,
            recorded_date=payload.recorded_date or date.today(),
        )
    )
    db.commit()
    db.refresh(profile)
    return profile


@router.get("/profiles/{user_id}/history", response_model=list[schemas.CreditScoreHistoryOut])
def get_history(
    user_id: str,
    db: Session = Depends(get_db),
    household_id: str = Depends(get_current_household_id),
):
    _get_household_user_or_404(db, household_id, user_id)
    return (
        db.query(models.CreditScoreHistory)
        .filter(models.CreditScoreHistory.user_id == user_id)
        .order_by(models.CreditScoreHistory.recorded_date)
        .all()
    )


@router.get("/profiles/{user_id}/recommendations", response_model=schemas.CreditRecommendationsOut)
def get_recommendations(
    user_id: str,
    db: Session = Depends(get_db),
    household_id: str = Depends(get_current_household_id),
):
    _get_household_user_or_404(db, household_id, user_id)
    profile = db.query(models.CreditProfile).filter(models.CreditProfile.user_id == user_id).first()
    if not profile:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "No credit profile on file yet for this person")
    return schemas.CreditRecommendationsOut(
        user_id=user_id,
        score=profile.score,
        score_band=score_band(profile.score),
        recommendations=build_recommendations(profile),
    )


@router.post("/borrowing-capacity", response_model=schemas.BorrowingCapacityOut)
def get_borrowing_capacity(
    payload: schemas.BorrowingCapacityRequest,
    db: Session = Depends(get_db),
    household_id: str = Depends(get_current_household_id),
):
    household = db.get(models.Household, household_id)
    annual_income = payload.annual_gross_income or float(household.annual_gross_income or 0)

    accounts = db.query(models.Account).filter(models.Account.household_id == household_id).all()
    debt_from_accounts = sum(float(a.monthly_payment or 0) for a in accounts if a.type in LIABILITY_TYPES)
    monthly_debt_payments = debt_from_accounts + payload.extra_monthly_debt

    if payload.cash_reserves_override is not None:
        cash_reserves = payload.cash_reserves_override
    else:
        cash_reserves = sum(
            float(a.current_balance) for a in accounts if a.type in (models.AccountType.checking, models.AccountType.savings)
        )

    profiles = db.query(models.CreditProfile).join(models.User).filter(models.User.household_id == household_id).all()
    best_score = max((p.score for p in profiles), default=None)

    result = compute_borrowing_capacity(
        annual_gross_income=annual_income,
        monthly_debt_payments=monthly_debt_payments,
        cash_reserves=cash_reserves,
        best_credit_score=best_score,
        interest_rate_pct=payload.interest_rate_pct,
        term_years=payload.term_years,
        down_payment_pct=payload.down_payment_pct,
    )
    return schemas.BorrowingCapacityOut(**result)

File: frontend/src/firebase/credit.ts
import { doc, getDoc, getDocs, query, serverTimestamp, where, writeBatch, type DocumentSnapshot } from "firebase/firestore";
import type {
  BorrowingCapacityOut,
  BorrowingCapacityRequest,
  CreditBureau,
  CreditProfile,
  CreditRecommendationsOut,
  CreditScoreHistoryPoint,
} from "../types";
import { db } from "./app";
import { computeBorrowingCapacity } from "./borrowingCapacity";
import { buildRecommendations, scoreBand } from "./creditAdvisor";
import { toDateString } from "./dates";
import { isLiabilityAccount } from "./finance";
import { HouseholdApi } from "./household";
import { AccountsApi } from "./ledger";
import { commit, householdCollection, isoString, snapshotData } from "./session";

// Each member's latest credit snapshot is stored at credit_profiles/{their user ID}, and
// every update also adds an entry to credit_score_history for the score chart.

function toProfile(snap: DocumentSnapshot): CreditProfile {
  const d = snapshotData(snap);
  return {
    id: snap.id,
    user_id: snap.id,
    bureau: d.bureau,
    score: d.score,
    utilization_pct: d.utilization_pct ?? null,
    on_time_payment_pct: d.on_time_payment_pct ?? null,
    credit_age_years: d.credit_age_years ?? null,
    num_hard_inquiries_12mo: d.num_hard_inquiries_12mo ?? null,
    num_open_accounts: d.num_open_accounts ?? null,
    num_derogatory_marks: d.num_derogatory_marks ?? null,
    updated_at: isoString(d.updated_at),
  };
}

type ProfileInput = Partial<Omit<CreditProfile, "id" | "user_id" | "updated_at">> & { recorded_date?: string };

function inRange(value: number | null | undefined, min: number, max: number, integer: boolean): boolean {
  return value == null || (value >= min && value <= max && (!integer || Number.isInteger(value)));
}

function validate(payload: ProfileInput): void {
  if (payload.score == null || !inRange(payload.score, 300, 850, true)) {
    throw new Error("Score must be a whole number from 300 to 850");
  }
  if (!inRange(payload.utilization_pct, 0, 100, false) || !inRange(payload.on_time_payment_pct, 0, 100, false)) {
    throw new Error("Percentages must be between 0 and 100");
  }
  if (!inRange(payload.credit_age_years, 0, Infinity, false)) throw new Error("Credit age can't be negative");
  for (const count of [payload.num_hard_inquiries_12mo, payload.num_open_accounts, payload.num_derogatory_marks]) {
    if (!inRange(count, 0, Infinity, true)) throw new Error("Counts must be whole numbers of 0 or more");
  }
}

async function requireMember(userId: string): Promise<void> {
  const member = await getDoc(doc(householdCollection("members"), userId));
  if (!member.exists()) throw new Error("Household member not found");
}

async function getProfile(userId: string): Promise<CreditProfile | null> {
  const snap = await getDoc(doc(householdCollection("credit_profiles"), userId));
  return snap.exists() ? toProfile(snap) : null;
}

export const CreditApi = {
  listProfiles: async (): Promise<CreditProfile[]> => {
    const snap = await getDocs(householdCollection("credit_profiles"));
    return snap.docs.map(toProfile);
  },

  /**
   * Replaces a member's credit snapshot (factors left out are cleared) and records the score
   * in their history, dated `recorded_date` or today. Any member can update anyone's.
   */
  upsertProfile: async (userId: string, payload: ProfileInput): Promise<CreditProfile> => {
    validate(payload);
    await requireMember(userId);
    const bureau: CreditBureau = payload.bureau ?? "other";
    const ref = doc(householdCollection("credit_profiles"), userId);

    const batch = writeBatch(db);
    batch.set(ref, {
      bureau,
      score: payload.score,
      utilization_pct: payload.utilization_pct ?? null,
      on_time_payment_pct: payload.on_time_payment_pct ?? null,
      credit_age_years: payload.credit_age_years ?? null,
      num_hard_inquiries_12mo: payload.num_hard_inquiries_12mo ?? null,
      num_open_accounts: payload.num_open_accounts ?? null,
      num_derogatory_marks: payload.num_derogatory_marks ?? null,
      updated_at: serverTimestamp(),
    });
    batch.set(doc(householdCollection("credit_score_history")), {
      user_id: userId,
      bureau,
      score: payload.score,
      recorded_date: payload.recorded_date || toDateString(new Date()),
      created_at: serverTimestamp(),
    });
    await commit(batch);
    return toProfile(await getDoc(ref));
  },

  /** Oldest first. */
  history: async (userId: string): Promise<CreditScoreHistoryPoint[]> => {
    await requireMember(userId);
    const snap = await getDocs(query(householdCollection("credit_score_history"), where("user_id", "==", userId)));
    return snap.docs
      .map((d) => ({ id: d.id, bureau: d.data().bureau, score: d.data().score, recorded_date: d.data().recorded_date }))
      .sort((a, b) => a.recorded_date.localeCompare(b.recorded_date));
  },

  recommendations: async (userId: string): Promise<CreditRecommendationsOut> => {
    await requireMember(userId);
    const profile = await getProfile(userId);
    if (!profile) throw new Error("No credit profile on file yet for this person");
    return {
      user_id: userId,
      score: profile.score,
      score_band: scoreBand(profile.score),
      recommendations: buildRecommendations(profile),
    };
  },

  /**
   * Estimates what the household could borrow, from its income (unless overridden), the
   * minimum payments on its credit card and loan accounts plus any extra debt, its checking
   * and savings balances (unless overridden), and the best credit score among its members.
   */
  borrowingCapacity: async (payload: BorrowingCapacityRequest = {}): Promise<BorrowingCapacityOut> => {
    const [household, accounts, profiles] = await Promise.all([HouseholdApi.get(), AccountsApi.list(), CreditApi.listProfiles()]);

    const debtFromAccounts = accounts
      .filter((a) => isLiabilityAccount(a.type))
      .reduce((sum, a) => sum + (a.monthly_payment ?? 0), 0);
    const liquidBalances = accounts
      .filter((a) => a.type === "checking" || a.type === "savings")
      .reduce((sum, a) => sum + a.current_balance, 0);

    return computeBorrowingCapacity({
      annualGrossIncome: payload.annual_gross_income || household.annual_gross_income || 0,
      monthlyDebtPayments: debtFromAccounts + (payload.extra_monthly_debt ?? 0),
      cashReserves: payload.cash_reserves_override ?? liquidBalances,
      bestCreditScore: profiles.length ? Math.max(...profiles.map((p) => p.score)) : null,
      interestRatePct: payload.interest_rate_pct ?? 6.5,
      termYears: payload.term_years ?? 30,
      downPaymentPct: payload.down_payment_pct ?? 20,
    });
  },
};
